### DISCUSSION TITLE
Deleting a non-existent layer still results in processing

### DISCUSSION THREAD

**Discussion by Jmuccigr** [ImageMagick/ImageMagick#5252] [Help]
I'm cleaning up some tiff output from a scanner that includes a preview layer. I'm just doing a simple

`mogrify -delete 1 filename.tiff`

Oddly IM will process the file even when there's no such layer. I can wrap this in a script to check for the layer, instead of just issuing the command directly, but shouldn't IM see that there's no second layer and simply leave the file alone?

PS the command is slightly more complex as I leave the tiff compression as-is.

**Comment by snibgo** [comment]
As a general principle, if an operation has no effect on your images, then IM won't do anything as a result of that operation, and won't warn you that it has no effect.

"mogrify" will read the files and re-write them. It does this even if the images have not changed as a result of any operations.

**Reply by Jmuccigr** [reply]
> As a general principle, if an operation has no effect on your images, then IM won't do anything as a result of that operation, and won't warn you that it has no effect.
> 
> "mogrify" will read the files and re-write them. It does this even if the images have not changed as a result of any operations.

So the general principle there doesn't apply to `mogrify` then?

I guess it's odd to me that mogrify will edit the file even when the thing it's acting on (a second frame) isn't there. I'd have expected some kind of error.

**Comment by snibgo** [comment]
The general principle applies to "mogrify". If the image (in IM's memory) does not change, that's not a problem. The operation in question is "-delete 1". When you have only one image, then that operation does nothing. It isn't an error because it doesn't prevent other operations from working. As with many no-op operations, IM doesn't even warn you.

The program, "magick" or "magick mogrify" will continue working, reading and writing whatever files you have told it to.

Warning when an operation doesn't change an image or the image list might be useful, but it would be expensive in program complexity, memory and time.

**Reply by Jmuccigr** [reply]
This issue I have is that IM processes the file even though it’s got nothing to work on. I’d expect it to see no such frame and just leave the file alone. Instead it “does something” to the file. I don’t mind the lack of warning, but the processing is a waste.

**Comment by snibgo** [comment]
Try:
```
mogrify filename.tiff
```
This will read filename.tiff, storing the images in memory. Then it will write those images to the same file, overwriting it. It is equivalent to:
```
magick filename.tiff filename.tiff
```
This might change the file greatly. For example, if the input is compressed with JPEG compression, the output won't be compressed at all.

Consider also:
```
magick filename.tiff tiff:- | someOtherProgram - options
```
We expect that magick will always send a tiff file to the pipe.

I agree that it might be useful if IM could kept a track of whether images change, and optionally suppress the file writing when there are no changes, but I think the cost of doing this would outweigh any benefits.

**Reply by Jmuccigr** [reply]
> I agree that it might be useful if IM could kept a track of whether images change, and optionally suppress the file writing when there are no changes, but I think the cost of doing this would outweigh any benefits.

I think generally that might be the case, but there are situations (like this one) where some simple logic would provide major benefits.

For example, before acting on layer N, IM could check whether layer N exists.

Obviously I'm not a developer here, so I could be totally off base, but I just wanted to point out what I think is odd behavior.

PS Regarding tiff compression, I've learned my lesson on that one and use `-define tiff:preserve-compression=true` to prevent those kinds of changes.

Anyway, I appreciate the conversation. Thanks.

**Reply by fmw42** [reply]
If you need to check whether you have one or more layers, then do that as a script. Check the number of layers and then use an IF test to decide to process or not. Specific tests should be done by the user as a script. Imagemagick should not have to predict what kinds of tests the user wants to do.

Every functions has a "null" operation, such as -rotate 0, which in principle could be trapped. But that would be a lot of effort for the staff to do. If someone wants to post all those code changes, it might be accepted.

**Reply by Jmuccigr** [reply]
> If you need to check whether you have one or more layers, then do that as a script. Check the number of layers and then use an IF test to decide to process or not. Specific tests should be done by the user as a script. Imagemagick should not have to predict what kinds of tests the user wants to do.
> 
> Every functions has a "null" operation, such as -rotate 0, which in principle could be trapped. But that would be a lot of effort for the staff to do. If someone wants to post all those code changes, it might be accepted.

Yeah, this is exactly what I ended up doing with a bash script. I get the part about "null" operations in a function, but it strikes me that this is slightly different, since the command is asking IM to act on a part of the image that doesn't exist, but perhaps in the code it isn't that much different.

Anyway, I appreciate the explanations.
